core/src/test/java/org/apache/calcite/test/SqlFunctionsTest.java

```diff
@@ -172,6 +173,58 @@ static <E> List<E> list() {
     assertThat(concat(null, "b"), is("nullb"));
   }
 
+  @Test void testSubString() {
+    // str vs single param
+    assertThat(substring("string", -1), is("string"));
+    assertThat(substring("string", -1L), is("string"));
+    assertThat(substring("string", 2), is("tring"));
+    assertThat(substring("string", 2L), is("tring"));
+    assertThat(substring("string", Integer.MIN_VALUE), is("string"));
+    assertThat(substring("string", Long.MIN_VALUE), is("string"));
+    assertThat(substring("string", Integer.MIN_VALUE + 10), is("string"));
+    assertThat(substring("string", Integer.MAX_VALUE), is(""));
+    assertThat(substring("string", Long.MAX_VALUE), is(""));
+    assertThat(substring("string", Integer.MAX_VALUE - 10), is(""));
+    assertThat(substring("string", Integer.MIN_VALUE - 10L), is("string"));
+    assertThat(substring("string", Integer.MAX_VALUE + 10L), is(""));
+
+    // str vs multi params
+    assertThat(substring("string", -1, 1), is(""));
```

Comment:
is this behavior documented?
it's not obvious whether clamping happens first or the index computation happens first.

Reply:
this equals to PG, what behavior are you talking about ?

Reply:
As I said, one can imagine that negative s is first clamped to 0 and then the function is evaluated with 0 and an unchanged l.
Does the reference.md file document this behavior?
It's also not obvious to me that substring's behavior is defined by Postgres.
According to https://calcite.apache.org/docs/reference.html#character-string-operators-and-functions it is a standard function.
We have seen many cases where various SQL dialects treat corner cases differently for function implementations (but this may be out of scope for the current PR).

Reply:
i understand what you are talking about but i check sql standard, there is nothing about concrete start position definition specified... if i need to fix smth here - plz let me know.

Reply:
The documentation is a bit sparse, but maybe you could improve it a bit describing what happens for negative values

Reply:
Fixed, if it`s not enough or clear can you suggest or fix it yourself, i`m not a native english speaking guy, thanks